syntax = "proto3";



message NoInputParameter {
}

message DescribeResponse
{
}


message ValueMessage
{
        message Attribute {
                enum Type {
                    DOUBLE  = 0;
                    INT    = 1;
                    BOOL   = 2;
                    STRING = 3;
                    LOCATION = 4;
                }
                string name = 1;
                Type type = 2;
                bool required = 3;
                string description = 4;
        }

        message DataModel {
                string name = 1;
                repeated Attribute attributes = 2;
                string description = 3;
        }

        message Location {
                string coordinate_system = 1; //e.g. latlon, latlonz
                string unit = 2;
                repeated double v = 3;
        }


        string typecode = 1;

// LOCATION is encoded into the doubles
// EMBEDDING is encoded into the doubles

        string s  = 2;
        float f   = 3;
        double d  = 4;
        int32 i32 = 5;
        int64 i64 = 6;
        bool b    = 7;

        DataModel dm = 8;
        Attribute a  = 9;
        Location l   = 10;

        repeated string v_s   = 102;
        repeated float v_f    = 103;
        repeated double v_d   = 104;
        repeated int32 v_i32  = 105;
        repeated int64 v_i64  = 106;
        repeated bool v_b     = 107;
        repeated Location v_l = 110;
}

message ConstructQueryConstraintObjectRequest
{
        string operator = 1;
        ValueMessage query_field_value = 2;
        string query_field_type = 3;
        string target_field_name = 4;
        string target_field_type = 5;
        string target_table_name = 6;
        string dap_name = 7;
        string node_name = 8;
}

message ConstructQueryObjectRequest
{
        string operator = 1;
        repeated ConstructQueryConstraintObjectRequest constraints = 2;
        repeated ConstructQueryObjectRequest children = 3;

// These two are used in internal representations.
        string node_name = 4;
        repeated string dap_names = 5;
}

message ConstructQueryMementoResponse
{
        bool success = 1;
        bytes memento = 2;
        string node_name = 3;
}

message Identifier
{
        bytes core = 1;
        bytes agent = 2;
        float score = 3;
        string uri = 4;
}

message DapExecute {
    ConstructQueryMementoResponse query_memento = 1;
    IdentifierSequence input_idents = 2;
}

message Successfulness
{
        bool success = 1;
        int32 errorcode = 2;
        repeated string narrative = 3;
}

message IdentifierSequence
{
        bool originator = 1;
        repeated Identifier identifiers = 2;
        Successfulness status = 3;
}

